cuda_core/cuda/core/experimental/_linker.py

```diff
@@ -0,0 +1,309 @@
+        if self.no_cache is not None:
+            self.formatted_options.append("-no-cache")
+
+
+class Linker:
+    """
+    Linker class for managing the linking of object codes with specified options.
+
+    Parameters
+    ----------
+    object_codes : ObjectCode
+        One or more ObjectCode objects to be linked.
+    options : LinkerOptions, optional
+        Options for the linker. If not provided, default options will be used.
+
+    Attributes
+    ----------
+    _options : LinkerOptions
+        The options used for the linker.
+    _handle : handle
+        The handle to the linker created by nvjitlink.
+
+    Methods
+    -------
+    _add_code_object(object_code)
+        Adds an object code to the linker.
+    close()
+        Closes the linker and releases resources.
+    """
+
+    class _MembersNeededForFinalize:
+        __slots__ = ("handle",)
+
+        def __init__(self, program_obj, handle):
+            self.handle = handle
+            weakref.finalize(program_obj, self.close)
+
+        def close(self):
+            if self.handle is not None:
+                nvjitlink.destroy(self.handle)
+                self.handle = None
+
+    __slots__ = ("__weakref__", "_mnff", "_options")
+
+    def __init__(self, *object_codes: ObjectCode, options: LinkerOptions = None):
+        self._options = options = check_or_create_options(LinkerOptions, options, "Linker options")
+        self._mnff.handle = nvjitlink.create(len(options.formatted_options), options.formatted_options)
+        self._mnff = Linker._MembersNeededForFinalize(self, None)
+
+        if len(object_codes) == 0:
+            raise ValueError("At least one ObjectCode object must be provided")
+
+        for code in object_codes:
+            assert isinstance(code, ObjectCode)
+            self._add_code_object(code)
+
+        weakref.finalize(self, self.close)
+
+    def _add_code_object(self, object_code: ObjectCode):
+        data = object_code._module
+        assert isinstance(data, bytes)
+        nvjitlink.add_data(
+            self._mnff.handle,
+            self._input_type_from_code_type(object_code._code_type),
+            data,
+            len(data),
+            f"{object_code._handle}_{object_code._code_type}",
+        )
+
+    def link(self, target_type) -> ObjectCode:
+        nvjitlink.complete(self._mnff.handle)
+        if target_type not in ("cubin", "ptx"):
+            raise ValueError(f"Unsupported target type: {target_type}")
+        code = None
+        if target_type == "cubin":
+            cubin_size = nvjitlink.get_linked_cubin_size(self._mnff.handle)
+            code = bytearray(cubin_size)
+            nvjitlink.get_linked_cubin(self._mnff.handle, code)
+        else:
+            ptx_size = nvjitlink.get_linked_ptx_size(self._mnff.handle)
+            code = bytearray(ptx_size)
+            nvjitlink.get_linked_ptx(self._mnff.handle, code)

```

Comment:
I'd do it like this, although it only saves a line, mainly because it's more obvious that the code is essentially the same for both `target_type`s, except for the difference in function names:

```python
        get_linked = {                                                          
            "cubin": (nvjitlink.get_linked_cubin_size, nvjitlink.get_linked_cubin),
            "ptx": (nvjitlink.get_linked_ptx_size, nvjitlink.get_linked_ptx),   
        }.get(target_type)                                                      
        if get_linked is None:                                                  
            raise ValueError(f"Unsupported target type: {target_type}")         
        get_size, get_code = get_linked                                         
        code_size = get_size(self._mnff.handle)                                 
        code = bytearray(code_size)                                             
        get_code(self._mnff.handle, code)                                       
```

Reply:
I like that revision